proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+```
+
+The `receiver` and `Length` expressions will be spilled as appropriate to ensure the side effects are only executed 
+once. For example:
+
+``` csharp
+class Collection {
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int this[int index] => _array[index];
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        int i = Get()[^1];
+        Console.WriteLine(i);
+    }
+}
+```
+
+This code will print "Get Length 3". 
+
+### Implicit Range support
+The language will provide an instance indexer member with a single parameter of type `Range` for types which meet the 
+following criteria:
+
+- The type is Countable.
+- The type has an accessible member named `Slice` which has two parameters of type `int`.
+- The type does not have an instance indexer which takes a single `Range` as the parameter.
+
+For such types the language will bind as if there is an index member of the form `T this[Range range]` where `T` is 
+the return type of the `Slice` method including any `ref` style annotations. The new member will also have matching
+accessibility with `Slice`. 
+
+When the `Range` based indexer is bound on an expression named `receiver` it will be lowered by converting the `Range` 
+expression into two values that are then passed to the `Slice` method. For discussion purposes lets use the example of 
+`receiver[expr]`.
+
+The first argument of `Slice` will be obtained by converting the typed expression in the following way:
+
+- When `expr` is in the form `expr1..expr2` (where `expr2` can be omitted) and `expr1` has type `int` then it will be
+emitted as `expr1`.
+- When `expr` is in the form `^expr1..expr2` (where `expr2` can be omitted) then it will be emitted as 
+`receiver.Length - expr1`
+- When `expr` is in the form `..expr2` (where `expr2` can be omitted) then it will be emitted as `0`
+- Else it will be emitted as `expr.Start.GetOffset(receiver.Length)` 
+
+This value will be re-used in the calculation of the second `Slice` argument. When doing so it will be referred to as
+`start`. The second argument of `Slice` will be obtained by converting the range typed expression in the following way:
+
+- When `expr` is in the form `expr1..expr2` (where `expr1` can be omitted) and `expr2` has type `int` then it will be 
+emitted as `expr2 - start`.
+- When `expr` is in the form `expr1..^expr2` (where `expr1` can be omitted) then it will be emitted as 
+`(receiver.Length - expr2) - start`
+- When `expr` is in the form `expr1..` (where `expr1` can be omitted) then it will be emitted as `receiver.Length`
+- Else it will be emitted as `expr.End.GetOffset(receiver.Length) - start` 
+
+The `receiver`, `Length` and `expr` expressions will be spilled as appropriate to ensure the side effects are only
```

Comment:
```suggestion
The `receiver`, `Length` and `expr` expressions will be spilled as appropriate to ensure any side effects are only
```
(To avoid assuming that there are always side effects)